packages/client/src/object/formatting/applyPropertyFormatter.test.ts

```diff
@@ -0,0 +1,133 @@
+  const mockClient = {} as MinimalClient;
+
+  // Single object definition with properties for testing
+  const OBJECT_DEF: FetchedObjectTypeDefinition = {
+    apiName: "TestObject",
+    type: "object",
+    primaryKeyApiName: "id",
+    titleProperty: "id",
+    rid: "ri.test.object",
+    displayName: "Test Object",
+    description: "A test object",
+    primaryKeyType: "string",
+    icon: undefined,
+    visibility: "NORMAL",
+    pluralDisplayName: "Test Objects",
+    status: "ACTIVE",
+    interfaceMap: {},
+    inverseInterfaceMap: {},
+    links: {},
+    implements: [],
+    [InterfaceDefinitions]: {},
+    properties: {
+      id: {
+        type: "string",
+        nullable: false,
+        multiplicity: false,
+      },
+      stringPropertyNoFormatting: {
+        type: "string",
+        nullable: true,
+        multiplicity: false,
+      },
+      booleanPropertyCustomLabels: {
+        type: "boolean",
+        nullable: false,
+        multiplicity: false,
+        valueFormatting: {
+          type: "boolean",
+          valueIfTrue: "Active",
+          valueIfFalse: "Inactive",
+        },
+      },
+      nullableBooleanProperty: {
+        type: "boolean",
+        nullable: true,
+        multiplicity: false,
+        valueFormatting: {
+          type: "boolean",
+          valueIfTrue: "Yes",
+          valueIfFalse: "No",
+        },
+      },
+    },
+  };
+
+  // Default object data with properties populated
+  const DEFAULT_OBJECT_DATA: SimpleOsdkProperties = {
+    $apiName: "TestObject",
+    $primaryKey: "test-123",
+    $title: "Test Object",
+    $objectType: "TestObject",
+    id: "test-123",
+    stringPropertyNoFormatting: "Plain text",
+    booleanPropertyCustomLabels: true,
+    nullableBooleanProperty: undefined,
+  };
+
+  // Helper to create an OSDK object with optional data overrides
+  function getObject(dataOverrides?: Partial<SimpleOsdkProperties>) {
+    const objectData = { ...DEFAULT_OBJECT_DATA, ...dataOverrides };
+    return createOsdkObject(mockClient, OBJECT_DEF, objectData);
+  }
+
+  it("formats boolean true with custom label", () => {
+    const obj = getObject();
+    const formatted = obj.$__EXPERIMENTAL__NOT_SUPPORTED_YET__getFormattedValue(
+      "booleanPropertyCustomLabels",
+    );
+
+    expect(formatted).toBe("Active");
+  });
+
+  it("formats boolean false with custom label", () => {
+    const obj = getObject({ booleanPropertyCustomLabels: false });
+    const formatted = obj.$__EXPERIMENTAL__NOT_SUPPORTED_YET__getFormattedValue(
+      "booleanPropertyCustomLabels",
+    );
+
+    expect(formatted).toBe("Inactive");
+  });
+
+  it("returns undefined for property without formatting rules", () => {
+    const obj = getObject();
+    const formatted = obj.$__EXPERIMENTAL__NOT_SUPPORTED_YET__getFormattedValue(
+      "stringPropertyNoFormatting",
+    );
+
+    expect(formatted).toBeUndefined();
+  });
+
+  it("returns undefined for null values", () => {
```

Comment:
just a small point, we call it `null` here but it's `undefined` above -- do we actually care to test `null` specifically or is all absence treated the same?